### PR TITLE
[move] Migrate most move-ir-compiler unit_tests

### PR DESCRIPTION
**Pull request by modocache**
This commit migrates most of the move-ir-compiler unit_tests, from Rust unit tests (which use Rust strings to represent Move IR programs), to the transactional test runner.

Two sets of tests remain as Rust unit tests:

1. Tests of the VM control flow graph. As far as I can tell, these should either be moved out of the move-ir-compiler test suite, or they require a transactional test runner that can verify the control-flow graph of a Move program, or they should be deleted. In any case, I'll do this in a separate commit.
2. Tests that use Rust to programmatically generate hundreds of functions and locals. Maybe these out to remain as Rust unit tests.

All tests other than these 2 kinds have been migrated to the transactional test runner. In doing so, I also categorized and documented the tests.

1. Tests of whether some text can be parsed as Move IR have been moved to a `parsing` directory.
2. Tests of whether bytecode can be generated from some parsed Move IR have been moved to a `bytecode-generation` directory.
3. Tests of the specification language have been moved to a `specification` directory.

Within each of these categories, tests of the same concept across several files have been combined into a single file. For example, all tests of how comments are parsed are now in `parsing/comments.mvir`.

## Motivation

Rust unit tests of the Move IR compiler have several downsides:

* The test assertions are verbose and annoying to write: "there ought to be 2 branch instructions," "there ought to be 4 load instructions," etc. It's much easier to make expectations against textual bytecode.
* Modifying the strings representing the Move IR programs requires re-compilation of the Rust test module that defines the string, which makes the test cycle slower.
* I frequently grep for Move/Move IR patterns in this repository by searching files that end in `.move` or `.mvir`. But these Move IR programs are defined as Rust strings in files that end with `.rs.`.
* I use the move-analyzer syntax highlighter for Move when viewing `.mvir` files (it works pretty well!). But most editors cannot syntax-highlight the Rust strings in these unit tests.

## Related PRs

The test runner's `print-bytecode` task can now disassemble compiled scripts, thanks to #9941.

### CODE REVIEW

File: language/move-ir-compiler/src/unit_tests/function_tests.rs [hunk 1]
```diff
@@ -87,25 +49,3 @@ fn compile_module_with_large_frame() {
...
-            label b0:
-                Self.foo();
-                return;
-            }
-        }
-        ",
-    );
-    let compiled_module_res = compile_module_string(&code);
-    assert!(compiled_module_res.is_ok());
-}
```
Comment (modocache) [inline language/move-ir-compiler/src/unit_tests/function_tests.rs:111]: These tests have moved to `language/move-ir-compiler/transactional-tests/tests/bytecode-generation/declarations/function.mvir`.

File: language/move-ir-compiler/src/unit_tests/serializer_tests.rs [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-    println!("binary[96]: {:?}", binary.get(96));
-    println!("binary[128]: {:?}", binary.get(128));
-    println!("binary[75]: {:?}", binary.get(75));
-    println!("binary[77]: {:?}", binary.get(77));
-    println!("binary[80]: {:?}", binary.get(80));
-    println!("binary[83]: {:?}", binary.get(83));
-    println!("binary[85]: {:?}", binary.get(85));
-    println!("binary[97]: {:?}", binary.get(97));
-    println!("binary[129]: {:?}", binary.get(129));
-    // println!("SCRIPT:\n{}", compiled_script);
```
Comment (modocache) [inline language/move-ir-compiler/src/unit_tests/serializer_tests.rs:37]: All of these print statements output `None`. The serializer is tested in many other crates in the project, including move-vm. And, in some ways, the expectation files in the move-ir-compiler-transactional-tests now test the bytecode output of Move IR function bodies better than this file used to. So, I think it's a win to have deleted this here.

Comment (vgao1996) [reply language/move-ir-compiler/src/unit_tests/serializer_tests.rs:37]: Yeah, this one does look a bit funky... Good job removing it!

File: language/move-ir-compiler/transactional-tests/tests/parsing/structs.mvir [hunk 3]
```diff
@@ -0,0 +1,13 @@
+//# print-bytecode --input=module
+// A struct declaration cannot contain function declarations (like the syntax for declaring
+// "methods" in other languages).
+module 0x3d8.M {
```
Comment (vgao1996) [inline language/move-ir-compiler/transactional-tests/tests/parsing/structs.mvir:4]: It seems like you've been using different addresses at different places. Just curious: is there any specific reason you're doing this? They seem to follow some sort of pattern...

**Review by vgao1996 — APPROVED** [review]
LGTM and I like that you've split everything into categories, clarifying what they are really testing.

File: language/move-ir-compiler/transactional-tests/tests/parsing/structs.mvir (hunk 3 above)
Comment (modocache) [reply language/move-ir-compiler/transactional-tests/tests/parsing/structs.mvir:4]: No meaningful pattern in terms of program semantics, but they do follow a pattern: dice combinations for Dungeons and Dragons-style tabletop games: this one is "three 8-sided dice," others are "two 20-sided dice" 🤓

**Pull request merged**